rfcs/text/0355-graph-non-primitive-data-types.md

```diff
@@ -0,0 +1,266 @@
+
+Ideally, data types and their constraints should be **composable**, so that a non-primitive data type can be defined as a combination of other data types. This will allow for expressing more complex data types, and enable the reuse of constraints across data types. Data types composition will allow users to _reduce_ and _expand_ value-spaces.
+
+For example, a user wants to model `color` as a property type within the type system. They want to represent a triple of numbers from 0 - 255. This is not possible in the current type system, but with composable non-primitive data types with constraints, it would be possible to define a `RGB Color` data type that consists of a list of three `Positive Integer` data types that are constrained to be between 0 and 255. `color` could make use of `RGB Color` to obtain the semantics the user seeks.
+
+The rationale for introducing **non-primitive** data type is therefore to allow for the type system to further constrain the value-space of its data types, in a reusable manner and allow for flexibility of defining more complex data types.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Non-primitive data types are data types that further constrain or extend an existing data type (which can be either non-primitive or primitive). In other words, non-primitive data types further reduce or expand the _value space_ of existing data types.
+
+JSON Schema allows for various constraints to be defined on schemas, and they can be combined in a variety of ways. When multiple constraints are defined in a JSON Schema, validators will try to apply each one even if they define unsatisfiable types. The Block Protocol type system will allow for the definition of non-primitive data types by combining these constraints in a similar way.
+
+To provide a way for these constraints to be composable and re-usable, the Block Protocol type system will allow for data types to _inherit_ from other data types. This will allow for data type constraints to be reusable and composable among non-primitive data types. This inheritance will lean on how the `allOf` keyword works in JSON Schema. All data types referenced in the `allOf` array of a data type schema will provide value-space constraints to the inheritor.
+
+A second kind of composability of data types will also be possible by allowing data types to be defined as choices between data types. Conceptually this is similar to JSON Schema's `oneOf` keyword where a schema can take on the shape of _one of_ a list of schemas. This composability, instead of constraining the value-space, will allow data types to expand their value-space.
+
+Lastly, we will allow for data types to compose through the use of lists. This means that data types can be of `{ type: "array" }` and use the `items` keyword to define their value-space(s).
```

Comment:
The specific reason why data types can be arrays if they can already be modeled using property types? This introduces some redundancy and I imagine confusion, as in: do I construct arrays in my `property-type` or `data-type`.